src/Abstract/Interpreter/Caching.hs

```diff
@@ -1,95 +1,125 @@
 import Data.Functor.Classes
 import qualified Data.Map as Map
 import Data.Maybe
-import Data.Proxy
-import qualified Data.Set as Set
+import Data.Pointed
+import Data.Semigroup
 import Data.Text.Prettyprint.Doc
 
-newtype Cache l t v = Cache { unCache :: Map.Map (Configuration l t v) (Set.Set (v, Store l v)) }
-  deriving (Monoid)
+newtype Cache l t v = Cache { unCache :: Map.Map (Configuration l t v) (Set (v, Store l v)) }
 
-cacheLookup :: (Ord t, Ord v, Address l) => Configuration l t v -> Cache l t v -> Maybe (Set.Set (v, Store l v))
-cacheLookup = (. unCache) . Map.lookup
+deriving instance (Ord l, Ord t, Ord v, Ord1 (Cell l)) => Monoid (Cache l t v)
 
-cacheSet :: (Ord t, Ord v, Address l) => Configuration l t v -> Set.Set (v, Store l v) -> Cache l t v -> Cache l t v
+cacheLookup :: (Ord l, Ord t, Ord v, Ord1 (Cell l)) => Configuration l t v -> Cache l t v -> Set (v, Store l v)
+cacheLookup key = fromMaybe mempty . Map.lookup key . unCache
+
+cacheSet :: (Ord l, Ord t, Ord v, Ord1 (Cell l)) => Configuration l t v -> Set (v, Store l v) -> Cache l t v -> Cache l t v
 cacheSet = (((Cache .) . (. unCache)) .) . Map.insert
 
-cacheInsert :: (Ord t, Ord v, Address l) => Configuration l t v -> (v, Store l v) -> Cache l t v -> Cache l t v
-cacheInsert = (((Cache .) . (. unCache)) .) . (. Set.singleton) . Map.insertWith (<>)
+cacheInsert :: (Ord l, Ord t, Ord v, Ord1 (Cell l)) => Configuration l t v -> (v, Store l v) -> Cache l t v -> Cache l t v
+cacheInsert = (((Cache .) . (. unCache)) .) . (. point) . Map.insertWith (<>)
+
+
+type CachingInterpreter l t v = '[Reader (Environment (Address l v)), Failure, NonDetEff, State (Store l v), Reader (Cache l t v), State (Cache l t v)]
+
+type CachingResult l t v = Final (CachingInterpreter l t v) v
+
+type MonadCachingInterpreter l t v m = (MonadEnv (Address l v) m, MonadStore l v m, MonadCacheIn l t v m, MonadCacheOut l t v m, Alternative m)
+
+
+class Monad m => MonadCacheIn l t v m where
+  askCache :: m (Cache l t v)
+  localCache :: (Cache l t v -> Cache l t v) -> m a -> m a
+
+instance Reader (Cache l t v) :< fs => MonadCacheIn l t v (Eff fs) where
+  askCache = ask
+  localCache = local
+
+
+class Monad m => MonadCacheOut l t v m where
+  getCache :: m (Cache l t v)
+  putCache :: Cache l t v -> m ()
+
+instance State (Cache l t v) :< fs => MonadCacheOut l t v (Eff fs) where
+  getCache = get
+  putCache = put
+
+modifyCache :: MonadCacheOut l t v m => (Cache l t v -> Cache l t v) -> m ()
+modifyCache f = fmap f getCache >>= putCache
 
 
-type CachingInterpreter l t v = Amb ': State (Cache l t v) ': Reader (Cache l t v) ': Interpreter l v
+class (Alternative m, Monad m) => MonadNonDet m where
+  collect :: Monoid b => (a -> b) -> m a -> m b
 
-type CachingResult l t v = (Either String ([] v, Cache l t v), Store l v)
+instance NonDetEff :< fs => MonadNonDet (Eff fs) where
+  collect f = interpose (pure . f) (\ m k -> case m of
+    MZero -> pure mempty
+    MPlus -> mappend <$> k True <*> k False)
```

Comment:
`interpose` allows us to handle nondeterminism requests while still leaving the capability for nondeterminism in place. We handle them by collecting them into a `Monoid`al result, essentially forcing the eager handling of them, rather than allowing them to propagate out to the caller.